File: ios/HtmlToImage.swift
import CoreImage
import CoreImage.CIFilterBuiltins
import ExpoModulesCore
import UIKit
import WebKit

class HtmlToImageConverter: NSObject, WKNavigationDelegate {
  private var webView: WKWebView
  private let width: CGFloat
  private let maxHeight: CGFloat
  private let completion: (Result<[Data], Error>) -> Void

  private lazy var opaqueFormat: UIGraphicsImageRendererFormat = {
    let format = UIGraphicsImageRendererFormat()
    format.opaque = true
    format.scale = 1.0
    format.preferredRange = .standard
    return format
  }()

  init(
    webView: WKWebView, width: CGFloat, maxHeight: CGFloat,
    completion: @escaping (Result<[Data], Error>) -> Void
  ) {
    self.webView = webView
    self.width = width
    self.maxHeight = maxHeight
    self.completion = completion
    super.init()
  }

  private func generateQrCode(text: String, size: CGFloat = 150, errorCorrection: String = "M") -> Data? {
    let data = text.data(using: .utf8)

    let filter = CIFilter.qrCodeGenerator()
    filter.message = data ?? Data()

    switch errorCorrection.uppercased() {
    case "L":
      filter.correctionLevel = "L"
    case "M":
      filter.correctionLevel = "M"
    case "Q":
      filter.correctionLevel = "Q"
    case "H":
      filter.correctionLevel = "H"
    default:
      filter.correctionLevel = "M"
    }

    guard let outputImage = filter.outputImage else { return nil }

    let scaleX = size / outputImage.extent.size.width
    let scaleY = size / outputImage.extent.size.height
    let transformedImage = outputImage.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))

    let context = CIContext()
    guard let cgImage = context.createCGImage(transformedImage, from: transformedImage.extent) else { return nil }
    let uiImage = UIImage(cgImage: cgImage)
    guard let imageData = uiImage.pngData() else { return nil }

    return imageData
  }

  private func generateBarcode(text: String, width: CGFloat = 200, height: CGFloat = 100) -> Data? {
    let data = text.data(using: .utf8)

    let filter = CIFilter.code128BarcodeGenerator()
    filter.quietSpace = 0
    filter.message = data ?? Data()

    guard let outputImage = filter.outputImage else { return nil }

    let scaleX = width / outputImage.extent.size.width
    let scaleY = height / outputImage.extent.size.height
    let transformedImage = outputImage.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))

    let context = CIContext()
    guard let cgImage = context.createCGImage(transformedImage, from: transformedImage.extent) else { return nil }
    let uiImage = UIImage(cgImage: cgImage)
    guard let imageData = uiImage.pngData() else { return nil }

    return imageData
  }

  static func renderHtmlToImages(config: [String: Any], html: String)
    async throws -> [Data]
  {
    let options = HtmlToImagesOptions(dictionary: config)
    let printerWidth: CGFloat = options.model == .model58 ? 384 : 576
    let maxHeight: CGFloat = options.maxHeightToBreak ?? 1600

    return try await withCheckedThrowingContinuation { continuation in
      // Create WebView on main thread
      DispatchQueue.main.async {
        let webView = WKWebView(frame: CGRect(x: 0, y: 0, width: printerWidth, height: 1))
        webView.isOpaque = true
        webView.backgroundColor = UIColor.white

        let delegate = HtmlToImageConverter(
          webView: webView,
          width: printerWidth,
          maxHeight: maxHeight,
          completion: { result in
            switch result {
            case .success(let imageDataArray):
              continuation.resume(returning: imageDataArray)
            case .failure(let error):
              continuation.resume(throwing: error)
            }
          }
        )

        webView.navigationDelegate = delegate
        objc_setAssociatedObject(webView, "delegate", delegate, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
        webView.loadHTMLString(html, baseURL: nil)
      }
    }
  }

  func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
    let convertScript = """
      function convertCodeElements() {
        const qrElements = document.querySelectorAll('[data-type="qrcode"]');
        const barcodeElements = document.querySelectorAll('[data-type="barcode"]');
        const allElements = [];
        
        qrElements.forEach((element, index) => {
          const text = element.getAttribute('data-text') || element.textContent || '';
          const size = element.getAttribute('data-size') || '150';
          const errorCorrection = element.getAttribute('data-error-correction') || 'M';
          
          allElements.push({
            type: 'qrcode',
            text: text,
            size: parseInt(size),
            errorCorrection: errorCorrection,
            elementId: 'qr-code-' + index
          });
          
          element.setAttribute('data-processing-id', 'qr-code-' + index);
        });
        
        barcodeElements.forEach((element, index) => {
          const text = element.getAttribute('data-text') || element.textContent || '';
          const width = element.getAttribute('data-width') || '200';
          const height = element.getAttribute('data-height') || `${width / 2}`;
          
          allElements.push({
            type: 'barcode',
            text: text,
            width: parseInt(width),
            height: parseInt(height),
            elementId: 'barcode-' + index
          });
          
          element.setAttribute('data-processing-id', 'barcode-' + index);
        });
        
        return allElements;
      }
      convertCodeElements();
      """

    webView.evaluateJavaScript(convertScript) { [weak self] (elementsData, error) in
      guard let self = self else { return }

      if let error = error {
        self.completion(.failure(error))
        return
      }

      guard let elements = elementsData as? [[String: Any]], !elements.isEmpty else {
        self.getContentHeightAndResize()
        return
      }

      self.processAllCodeElements(elements: elements)
    }
  }

  private func processAllCodeElements(elements: [[String: Any]]) {
    var processedElements: [String: String] = [:]

    for elementData in elements {
      guard let type = elementData["type"] as? String,
        let text = elementData["text"] as? String,
        let elementId = elementData["elementId"] as? String
      else {
        continue
      }

      var base64Image: String?
      if type == "qrcode" {
        guard let size = elementData["size"] as? Int else { continue }
        let errorCorrection = elementData["errorCorrection"] as? String ?? "M"
        let image = self.generateQrCode(text: text, size: CGFloat(size), errorCorrection: errorCorrection)
        base64Image = image?.base64EncodedString()
      } else if type == "barcode" {
        let width = elementData["width"] as? Int ?? 200
        let height = elementData["height"] as? Int ?? 100
        let image = self.generateBarcode(text: text, width: CGFloat(width), height: CGFloat(height))
        base64Image = image?.base64EncodedString()
      }

      if let imageBase64 = base64Image {
        processedElements[elementId] = imageBase64
      }
    }

    self.updateAllElements(processedElements: processedElements)
  }

  private func updateAllElements(processedElements: [String: String]) {
    let updateScript = """
      (function() {
        \(processedElements.map { (elementId, base64) in
        return """
        const element\(elementId.replacingOccurrences(of: "-", with: "_")) = document.querySelector('[data-processing-id="\(elementId)"]');
        if (element\(elementId.replacingOccurrences(of: "-", with: "_"))) {
          const img = document.createElement('img');
          img.src = 'data:image/png;base64,\(base64)';
          img.style.display = 'block';
          element\(elementId.replacingOccurrences(of: "-", with: "_")).innerHTML = '';
          element\(elementId.replacingOccurrences(of: "-", with: "_")).appendChild(img);
        }
        """
      }.joined(separator: "\n"))
      })();
      """

    webView.evaluateJavaScript(updateScript) { [weak self] (_, error) in
      guard let self = self else { return }

      if let error = error {
        self.completion(.failure(error))
        return
      }

      self.getContentHeightAndResize()
    }
  }

  private func getContentHeightAndResize() {
    webView.evaluateJavaScript("document.body.scrollHeight") { [weak self] (height, error) in
      guard let self = self, let contentHeight = height as? CGFloat else {
        self?.completion(
          .failure(
            NSError(
              domain: "WebViewCapture", code: 1,
              userInfo: [
                NSLocalizedDescriptionKey: "Failed to get content height"
              ])))
        return
      }

      // Resize WebView to exact content dimensions
      self.webView.frame = CGRect(x: 0, y: 0, width: self.width, height: contentHeight)

      // Take snapshot after resize
      self.takeSnapshot()
    }
  }

  private func sliceImage(_ image: UIImage, width: CGFloat, maxHeight: CGFloat) -> [Data] {
    var imageParts: [Data] = []
    let totalHeight = image.size.height
    var currentY: CGFloat = 0

    while currentY < totalHeight {
      let sliceHeight = min(maxHeight, totalHeight - currentY)

      let renderer = UIGraphicsImageRenderer(size: CGSize(width: width, height: sliceHeight), format: opaqueFormat)
      let sliceImage = renderer.image { context in
        UIColor.white.setFill()
        context.fill(CGRect(x: 0, y: 0, width: width, height: sliceHeight))
        image.draw(at: CGPoint(x: 0, y: -currentY))
      }

      if let pngData = sliceImage.pngData() {
        imageParts.append(pngData)
      }

      currentY += sliceHeight
    }

    return imageParts
  }

  private func takeSnapshot() {
    let config = WKSnapshotConfiguration()
    config.rect = webView.bounds

    webView.takeSnapshot(with: config) { [weak self] (image, error) in
      guard let self = self else { return }

      if let error = error {
        self.completion(.failure(error))
        return
      }

      guard let image = image else {
        self.completion(
          .failure(
            NSError(
              domain: "WebViewCapture", code: 2,
              userInfo: [NSLocalizedDescriptionKey: "Failed to create image"])
          ))
        return
      }

      // Process image in background
      DispatchQueue.global(qos: .userInitiated).async {
        let targetWidth: CGFloat = self.width
        let targetHeight = image.size.height * (targetWidth / image.size.width)

        let renderer = UIGraphicsImageRenderer(
          size: CGSize(width: targetWidth, height: targetHeight), format: self.opaqueFormat)
        let resizedImage = renderer.image { context in
          UIColor.white.setFill()
          context.fill(CGRect(x: 0, y: 0, width: targetWidth, height: targetHeight))
          image.draw(in: CGRect(x: 0, y: 0, width: targetWidth, height: targetHeight))
        }

        let imageParts = self.sliceImage(resizedImage, width: targetWidth, maxHeight: self.maxHeight)

        DispatchQueue.main.async {
          if imageParts.isEmpty {
            self.completion(
              .failure(
                NSError(
                  domain: "WebViewCapture", code: 4,
                  userInfo: [
                    NSLocalizedDescriptionKey: "Failed to create image slices"
                  ])))
            return
          }

          self.completion(.success(imageParts))
        }
      }
    }
  }
}

File: android/src/main/java/expo/modules/escpos/HtmlToImage.kt
package expo.modules.escpos

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.webkit.WebView
import android.webkit.WebViewClient
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.MultiFormatWriter
import com.google.zxing.common.BitMatrix
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import kotlin.math.min

object HtmlToImageConverter {
    private val coroutineScope = CoroutineScope(Dispatchers.Default)

    private fun generateQrCode(text: String, size: Int = 150, errorCorrection: String = "M"): ByteArray? {
        return try {
            val hints = hashMapOf<EncodeHintType, Any>()
            hints[EncodeHintType.MARGIN] = 0

            when (errorCorrection.uppercase()) {
                "L" -> hints[EncodeHintType.ERROR_CORRECTION] = ErrorCorrectionLevel.L
                "M" -> hints[EncodeHintType.ERROR_CORRECTION] = ErrorCorrectionLevel.M
                "Q" -> hints[EncodeHintType.ERROR_CORRECTION] = ErrorCorrectionLevel.Q
                "H" -> hints[EncodeHintType.ERROR_CORRECTION] = ErrorCorrectionLevel.H
                else -> hints[EncodeHintType.ERROR_CORRECTION] = ErrorCorrectionLevel.M
            }

            val writer = MultiFormatWriter()
            val bitMatrix = writer.encode(text, BarcodeFormat.QR_CODE, size, size, hints)
            
            val bitmap = createBitmapFromBitMatrix(bitMatrix)
            bitmapToByteArray(bitmap)
        } catch (e: Exception) {
            null
        }
    }
    
    private fun generateBarcode(text: String, width: Int = 200, height: Int = 100): ByteArray? {
        return try {
            val hints = hashMapOf<EncodeHintType, Any>()
            hints[EncodeHintType.MARGIN] = 0

            val writer = MultiFormatWriter()
            val bitMatrix = writer.encode(text, BarcodeFormat.CODE_128, width, height, hints)
            
            val bitmap = createBitmapFromBitMatrix(bitMatrix)
            bitmapToByteArray(bitmap)
        } catch (e: Exception) {
            null
        }
    }
    
    private fun createBitmapFromBitMatrix(bitMatrix: BitMatrix): Bitmap {
        val width = bitMatrix.width
        val height = bitMatrix.height
        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        
        for (x in 0 until width) {
            for (y in 0 until height) {
                bitmap.setPixel(x, y, if (bitMatrix[x, y]) Color.BLACK else Color.WHITE)
            }
        }
        
        return bitmap
    }

    private fun bitmapToByteArray(bitmap: Bitmap): ByteArray {
        val outputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
        val byteArray = outputStream.toByteArray()
        outputStream.close()
        bitmap.recycle()
        return byteArray
    }

    @SuppressLint("SetJavaScriptEnabled")
    fun renderHtmlToImages(
        context: Context,
        html: String,
        width: Int,
        maxHeightToBreak: Int,
        callback: (Result<List<ByteArray>>) -> Unit
    ) {
        // Must run WebView on main thread
        Handler(Looper.getMainLooper()).post {
            try {
                val webView = WebView(context)
                webView.settings.javaScriptEnabled = true
                webView.settings.useWideViewPort = true
                webView.setBackgroundColor(Color.WHITE)

                // Set the initial size
                webView.layout(0, 0, width, 1)

                webView.webViewClient = object : WebViewClient() {
                    override fun onPageFinished(view: WebView, url: String) {
                        // First convert barcodes and QR codes to images
                        val convertScript = """
                        function convertCodeElements() {
                            const qrElements = document.querySelectorAll('[data-type="qrcode"]');
                            const barcodeElements = document.querySelectorAll('[data-type="barcode"]');
                            const allElements = [];
                            
                            qrElements.forEach((element, index) => {
                                const text = element.getAttribute('data-text') || element.textContent || '';
                                const size = element.getAttribute('data-size') || '150';
                                const errorCorrection = element.getAttribute('data-error-correction') || 'M';
                                
                                allElements.push({
                                    type: 'qrcode',
                                    text: text,
                                    size: parseInt(size),
                                    errorCorrection: errorCorrection,
                                    elementId: 'qr-code-' + index
                                });
                                
                                element.setAttribute('data-processing-id', 'qr-code-' + index);
                            });
                            
                            barcodeElements.forEach((element, index) => {
                                const text = element.getAttribute('data-text') || element.textContent || '';
                                const width = element.getAttribute('data-width') || '200';
                                const height = element.getAttribute('data-height') || `${'$'}{width / 2}`;
                                
                                allElements.push({
                                    type: 'barcode',
                                    text: text,
                                    width: parseInt(width),
                                    height: parseInt(height),
                                    elementId: 'barcode-' + index
                                });
                                
                                element.setAttribute('data-processing-id', 'barcode-' + index);
                            });
                            
                            return allElements;
                        }
                        convertCodeElements();
                        """.trimIndent()

                        webView.evaluateJavascript(convertScript) { elementsDataJson ->
                            // Process the elements data and generate images
                            coroutineScope.launch {
                                try {
                                    val processedElements = processCodeElements(elementsDataJson)

                                    withContext(Dispatchers.Main) {
                                        updateAllElements(webView, processedElements) {
                                            // After updating elements, get final content height
                                            getContentHeightAndResize(webView, width, maxHeightToBreak, callback)
                                        }
                                    }
                                } catch (e: Exception) {
                                    withContext(Dispatchers.Main) {
                                        webView.destroy()
                                    }
                                    callback(Result.failure(e))
                                }
                            }
                        }
                    }
                }

                // Load the HTML
                webView.loadDataWithBaseURL(null, html, "text/html", "UTF-8", null)
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }
    
    private suspend fun processCodeElements(elementsDataJson: String): Map<String, String> {
        return withContext(Dispatchers.Default) {
            val processedElements = mutableMapOf<String, String>()

            try {
                // Parse the JSON-like string (simplified parsing for the data structure we expect)
                val elementsData = parseElementsData(elementsDataJson)

                for (elementData in elementsData) {
                    val type = elementData["type"] as? String ?: continue
                    val text = elementData["text"] as? String ?: continue
                    val elementId = elementData["elementId"] as? String ?: continue
                    
                    val imageData = when (type) {
                        "qrcode" -> {
                            val size = (elementData["size"] as? Int) ?: 150
                            val errorCorrection = (elementData["errorCorrection"] as? String) ?: "M"
                            generateQrCode(text, size, errorCorrection)
                        }
                        "barcode" -> {
                            val width = (elementData["width"] as? Int) ?: 200
                            val height = (elementData["height"] as? Int) ?: 100
                            generateBarcode(text, width, height)
                        }
                        else -> null
                    }
                    
                    imageData?.let {
                        val base64String = Base64.encodeToString(it, Base64.NO_WRAP)
                        processedElements[elementId] = base64String
                    }
                }
            } catch (e: Exception) {
                // Log error but continue with empty map
            }
            
            processedElements
        }
    }
    
    private fun parseElementsData(jsonString: String): List<Map<String, Any>> {
        val elements = mutableListOf<Map<String, Any>>()
        
        val cleanJson = jsonString.trim().removeSurrounding("[", "]")
        
        if (cleanJson.isBlank()) return elements
        
        val objectStrings = cleanJson.split("},{").map {
            if (!it.startsWith("{")) "{$it" else it
        }.map {
            if (!it.endsWith("}")) "$it}" else it
        }
        
        for (objStr in objectStrings) {
            try {
                val element = mutableMapOf<String, Any>()
                
                // Extract key-value pairs
                val content = objStr.removeSurrounding("{", "}")
                val pairs = content.split(",")
                
                for (pair in pairs) {
                    val keyValue = pair.split(":")
                    if (keyValue.size == 2) {
                        val key = keyValue[0].trim().removeSurrounding("\"")
                        val value = keyValue[1].trim()
                        
                        when {
                            value.startsWith("\"") && value.endsWith("\"") -> {
                                element[key] = value.removeSurrounding("\"")
                            }
                            value.toIntOrNull() != null -> {
                                element[key] = value.toInt()
                            }
                            else -> {
                                element[key] = value.removeSurrounding("\"")
                            }
                        }
                    }
                }
                
                if (element.isNotEmpty()) {
                    elements.add(element)
                }
            } catch (e: Exception) {
                // Skip malformed objects
            }
        }
        
        return elements
    }
    
    private fun updateAllElements(webView: WebView, processedElements: Map<String, String>, onComplete: () -> Unit) {
        if (processedElements.isEmpty()) {
            onComplete()
            return
        }
        
        val updateScript = processedElements.map { (elementId, base64) ->
            val safeElementId = elementId.replace("-", "_")
            """
            const element$safeElementId = document.querySelector('[data-processing-id="$elementId"]');
            if (element$safeElementId) {
                const img = document.createElement('img');
                img.src = 'data:image/png;base64,$base64';
                img.style.display = 'block';
                element$safeElementId.innerHTML = '';
                element$safeElementId.appendChild(img);
            }
            """.trimIndent()
        }.joinToString("\n")
        
        webView.evaluateJavascript("(function() { $updateScript })();") {
            onComplete()
        }
    }
    
    private fun getContentHeightAndResize(
        webView: WebView, 
        width: Int, 
        maxHeightToBreak: Int, 
        callback: (Result<List<ByteArray>>) -> Unit
    ) {
        val density = webView.resources.displayMetrics.density
        val zoomPercentage = (1.0 / density * 100).toInt()
        val jsScript = """
            (function() {
                document.body.style.zoom = '$zoomPercentage%';
                return document.body.scrollHeight;
            })();
        """.trimIndent()

        webView.evaluateJavascript(jsScript) { heightString ->
            val height = heightString.toFloatOrNull()

            if (height != null) {
                val contentHeight = (height * density).toInt()

                try {
                    // Resize WebView to content dimensions
                    webView.layout(0, 0, width, contentHeight)

                    // Wait a bit more for rendering to complete after resize
                    Handler(Looper.getMainLooper()).postDelayed({
                        // Process bitmap in background using coroutines
                        coroutineScope.launch {
                            try {
                                val imageParts = withContext(Dispatchers.Default) {
                                    processBitmap(webView, width, maxHeightToBreak)
                                }

                                // Clean up WebView on main thread
                                withContext(Dispatchers.Main) {
                                    webView.destroy()
                                }

                                if (imageParts.isEmpty()) {
                                    callback(Result.failure(Exception("Failed to create image slices")))
                                } else {
                                    callback(Result.success(imageParts))
                                }
                            } catch (e: Exception) {
                                withContext(Dispatchers.Main) {
                                    webView.destroy()
                                }
                                callback(Result.failure(e))
                            }
                        }
                    }, 100) // Additional wait after resize
                } catch (e: Exception) {
                    webView.destroy()
                    callback(Result.failure(Exception("Failed to get content height: ${e.message}")))
                }
            }
        }
    }

    private suspend fun processBitmap(webView: WebView, width: Int, maxHeightToBreak: Int): List<ByteArray> {
        return withContext(Dispatchers.IO) {
            val bitmap = Bitmap.createBitmap(webView.width, webView.height, Bitmap.Config.ARGB_8888)
            val canvas = android.graphics.Canvas(bitmap)
            
            withContext(Dispatchers.Main) {
                webView.draw(canvas)
            }

            val imageParts = mutableListOf<ByteArray>()
            val totalHeight = bitmap.height
            var currentY = 0

            try {
                while (currentY < totalHeight) {
                    val sliceHeight = min(maxHeightToBreak, totalHeight - currentY)

                    // Create a bitmap for this slice
                    val sliceBitmap = Bitmap.createBitmap(
                        bitmap, 0, currentY, width, sliceHeight
                    )

                    // Convert to PNG
                    val outputStream = ByteArrayOutputStream()
                    sliceBitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
                    imageParts.add(outputStream.toByteArray())

                    // Clean up
                    sliceBitmap.recycle()
                    outputStream.close()

                    // Move to next slice
                    currentY += sliceHeight
                }
            } finally {
                // Clean up main bitmap
                bitmap.recycle()
            }

            imageParts
        }
    }
}
